File: src/main/java/io/shcm/shsupercm/fabric/stonecutter/version/VersionChecker.java
package io.shcm.shsupercm.fabric.stonecutter.version;

import io.shcm.shsupercm.fabric.stonecutter.processor.expression.ExpressionResult;

import java.util.function.Predicate;

public interface VersionChecker {
    Object parseVersion(String versionString) throws Exception;

    Predicate<Object> parseChecker(String predicateString) throws Exception;

    default ExpressionResult check(Object version, String predicateString) {
        try {
            return ExpressionResult.of(parseChecker(predicateString).test(version));
        } catch (Exception e) {
            return ExpressionResult.INVALID;
        }
    }
}

File: src/main/kotlin/dev/kikugie/stonecutter/version/VersionChecker.kt
package dev.kikugie.stonecutter.version

import java.util.function.Predicate

typealias Version = Any

interface VersionChecker {
    fun parseVersion(version: String): Version
    fun parsePredicate(predicate: String): Predicate<Version>
    fun test(version: Version, predicate: String): Boolean? = try {
        parsePredicate(predicate).test(version)
    } catch (e: Exception) {
        null
    }
}